services/src/pro/users/postgres_userdb.rs

```diff
@@ -449,12 +470,204 @@ where
+
+        let db_valid_until = tx
+            .query_one(
+                &stmt,
+                &[
+                    &session,
+                    &flat_tokens.access_token_value,
+                    &flat_tokens.access_token_nonce,
+                    &(oidc_tokens.expires_in.num_seconds() as f64),
+                    &flat_tokens.refresh_token_value,
+                    &flat_tokens.refresh_token_nonce,
+                ],
+            )
+            .await?
+            .get(0);
+
+        Ok(StoredOidcTokens {
+            oidc_tokens,
+            db_valid_until,
+        })
+    }
+
+    async fn refresh_tokens(
+        &self,
+        session: SessionId,
+        tx: &Transaction<'_>,
+    ) -> Result<StoredOidcTokens> {
+        let stmt = tx
+            .prepare(
+                "
+            SELECT
+                refresh_token,
+                refresh_token_encryption_nonce
+            FROM
+                oidc_session_tokens
+            WHERE session_id = $1 AND refresh_token IS NOT NULL;",
+            )
+            .await?;
+
+        let rows = tx.query_opt(&stmt, &[&session]).await?;
+
+        if let Some(refresh_string) = rows {
+            let string_field_and_nonce = MaybeEncryptedBytes {
+                value: refresh_string.get(0),
+                nonce: refresh_string.get(1),
+            };
+
+            let refresh_token = self
+                .oidc_manager()
+                .maybe_decrypt_refresh_token(string_field_and_nonce)?;
+
+            let oidc_manager = self.oidc_manager();
+
+            let oidc_tokens = oidc_manager
+                .get_client()
+                .await?
+                .refresh_token(refresh_token)
+                .await?;
+
+            let flat_tokens: FlatMaybeEncryptedOidcTokens = self
+                .oidc_manager()
+                .maybe_encrypt_tokens(&oidc_tokens)?
+                .into();
+
+            let update_session_tokens = tx.prepare("
+                UPDATE
+                    oidc_session_tokens
+                SET
+                    access_token = $2, access_token_encryption_nonce = $3, access_token_valid_until = CURRENT_TIMESTAMP + make_interval(secs:=$4), refresh_token = $5, refresh_token_encryption_nonce = $6
+                WHERE
+                    session_id = $1;",
+
+            ).await?;
+
+            let updated_token_rows = tx
+                .execute(
+                    &update_session_tokens,
+                    &[
+                        &session,
+                        &flat_tokens.access_token_value,
+                        &flat_tokens.access_token_nonce,
+                        &(oidc_tokens.expires_in.num_seconds() as f64),
+                        &flat_tokens.refresh_token_value,
+                        &flat_tokens.refresh_token_nonce,
+                    ],
+                )
+                .await?;
+
+            if updated_token_rows != 1 {
+                return Err(Error::InvalidSession);
```

Comment:
when can this happen and is there a more meaningful error (to log at least)?

Reply:
Since the same transactions first checks if this row exists, I don't think this can occur. ee8a2fc34b0071a07bfcd86d003e9da7adb57ff2